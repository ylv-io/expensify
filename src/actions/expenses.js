import uuid from 'uuid';
import database from '../firebase/firebase';

export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
});

export const startAddExpense = (expenseData = {}) => {
  return (dispatch, getState) => {
    const { uid } = getState().auth;
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData;
    const expense = { description, note, amount, createdAt };

    return database
      .ref(`users/${uid}/expenses`)
      .push(expense)
      .then((ref) => {
        dispatch(addExpense({
          id: ref.key,
          ...expense,
        }))
      });
  };
};

export const removeExpense = (
  { 
    id
  }
) => ({
  type: 'REMOVE_EXPENSE',
  id
});

export const startRemoveExpense = ({ id }) => (dispatch, getState) => {
  const { uid } = getState().auth;
  return database
      .ref(`users/${uid}/expenses/${id}`)
      .remove()
      .then(() => {
        dispatch(removeExpense({ id }));
      });
};

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

export const startEditExpense = (id, updates) => (dispatch, getState) => {
  const { uid } = getState().auth;
  return database
      .ref(`users/${uid}/expenses/${id}`)
      .update(updates)
      .then(() => {
        dispatch(editExpense(id, updates));
      });
};

// SET_EXPENSES
export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses,
});

export const startSetExpenses = () => (dispatch, getState) => {
  const { uid } = getState().auth;
  return database
      .ref(`users/${uid}/expenses`)
      .once('value')
      .then((snapshot) => {
        const expenses = Object.keys(snapshot.val() || {}).map(k => ({ id: k, ...snapshot.val()[k] }));
        dispatch(setExpenses(expenses));
      });
};