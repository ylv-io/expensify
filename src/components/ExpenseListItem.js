import React from 'react';
import moment from 'moment';
import numeral from 'numeral';

import { BrowserRouter, Route, Switch, Link, NavLink } from 'react-router-dom';

const ExpenseListItem = ({id, description, amount, createdAt}) => (
  <div>
    <h3>
      <Link to={`/edit/${id}`}>
        {description}
      </Link>
    </h3>
    <p>{numeral(amount / 100).format('$0,0.00')} - {moment(createdAt).format('MMM Do YYYY')}</p>
  </div>
);

export default ExpenseListItem;