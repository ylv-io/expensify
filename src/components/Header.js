import React from 'react';
import { BrowserRouter, Route, Switch, Link, NavLink } from 'react-router-dom';


const Header = () => (
  <header>
    <h1>Expensify</h1>
    <ul>
      <li>
        <NavLink to="/" activeClassName="is-active" exact={true}>home</NavLink>
      </li>
      <li>
        <NavLink to="/create" activeClassName="is-active" exact={true}>create</NavLink>
      </li>

    </ul>
  </header>
);

export default Header;
