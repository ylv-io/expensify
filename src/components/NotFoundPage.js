import React from 'react';
import { BrowserRouter, Route, Switch, Link, NavLink } from 'react-router-dom';

const NotFoundPage = () => (
  <div>
    404 – <Link to="/">home</Link>
  </div>
);

export default NotFoundPage;