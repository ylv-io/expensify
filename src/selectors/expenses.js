import moment from 'moment';

export default (expenses, {text, sortBy, startDate, endDate}) => {
  return expenses.filter((o) => {
    const createdAtMoment = moment(o.createdAt);
    const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
    const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
    const textMatch = !text || o.description.toLowerCase().includes(text.toLowerCase());

    return startDateMatch && endDateMatch && textMatch;
  }).sort((a, b) => {
    if (sortBy === 'date') {
      return a.createdAt < b.createdAt ? 1 : -1;
    }
    if (sortBy === 'amount') {
      return a.amount < b.amount ? 1 : -1;
    }
  });
};