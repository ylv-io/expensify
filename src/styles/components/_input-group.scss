.input-group {
  display: flex;
  margin-bottom: $m-size;
  flex-direction: column;

  @media (min-width: $desktop-breakpoint) {
    flex-direction: row;
    margin-bottom: $l-size;
  }
}

.input-group__item {
  margin-bottom: $s-size;

  @media (min-width: $desktop-breakpoint) {
    margin-bottom: 0;
    margin-right: $s-size;
  }
}
