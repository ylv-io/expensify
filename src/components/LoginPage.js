import React from 'react';
import { connect } from 'react-redux';
import { startLogin } from '../actions/auth';
import { dispatch } from 'rxjs/internal/observable/range';

export const LoginPage = ({ startLogin }) => (
  <div>
    <button onClick={startLogin}>Login</button>
  </div>
);

const mapDispatchToProps = (dispatch) => ({
  startLogin: () => dispatch(startLogin())
});

export default connect(undefined, mapDispatchToProps)(LoginPage);