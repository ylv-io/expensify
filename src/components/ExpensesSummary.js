import React from 'react';
import { connect } from 'react-redux';

import numeral from 'numeral';

import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';
import expensesTotal from '../selectors/expenses-total';

export const ExpensesSummary = (props) => (
  <div>
    <p>
      Viewing {props.expenses.length} expenses totalling {numeral(props.expensesTotal / 100).format('$0,0.00')}
    </p>
  </div>
);

const mapStateToProps = (state) => {
  const expenses = selectExpenses(state.expenses, state.filters);
  return {
    expenses: expenses,
    expensesTotal: expensesTotal(expenses),
  };
};

export default connect(mapStateToProps)(ExpensesSummary);