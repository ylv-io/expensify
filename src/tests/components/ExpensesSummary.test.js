import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';

test('should render ExpensesSummary with expenses', () => {
  const wrapper = shallow(<ExpensesSummary expenses={expenses} expensesTotal={23423} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render ExpensesSummary with no expenses', () => {
  const wrapper = shallow(<ExpensesSummary expenses={[]} expensesTotal={0} />);
  expect(wrapper).toMatchSnapshot();
});
