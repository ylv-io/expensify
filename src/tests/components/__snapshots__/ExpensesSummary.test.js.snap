// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesSummary with expenses 1`] = `
<div>
  <p>
    Viewing 
    3
     expenses totalling 
    $234.23
  </p>
</div>
`;

exports[`should render ExpensesSummary with no expenses 1`] = `
<div>
  <p>
    Viewing 
    0
     expenses totalling 
    $0.00
  </p>
</div>
`;
